### PR TITLE
Aurora Cluster support

### PR DESCRIPTION
**Pull request by sandrampeter**
## Background

Adding example for Aurora RDS cluster with 1 Reader and 1 Writer. TFE will use the Cluster Endpoint for connecting to database. Increased the private subnet created by default to 3 as Aurora RDS cluster requires min 3 Azs for deployment.

[JIRA](https://hashicorp.atlassian.net/browse/TF-14278)

### Test Configuration

Used Aurora PostgreSQL - version 16.2 (db.r5.xlarge)

### CODE REVIEW

File: modules/aurora_database_cluster/variables.tf [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+variable "cluster_members" {
+  description = "List of RDS Instances that are a part of this cluster"
+  type        = list(string)
+  default     = null
+}
+
+variable "replica_count" {
+  type        = string
+  default     = "1"
+  description = "Number of reader nodes to create.  If `replica_scale_enable` is `true`, the value of `replica_scale_min` is used instead."
```
Comment (nikolasrieble) [inline modules/aurora_database_cluster/variables.tf:98]: ❓ I could not find the variables `replica_scale_enable` and `replica_scale_min`. What does this description refer to?

File: modules/aurora_database_cluster/variables.tf [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+variable "single_instance_enabled" {
+  type        = string
+  default     = true
+  description = "Whether the database resources should be created"
```
Comment (nikolasrieble) [inline modules/aurora_database_cluster/variables.tf:104]: ❓ I am not sure I understand. If we set this to false, no database resources are created? In which case would one set to this false?

File: modules/aurora_database_cluster/variables.tf [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+variable "db_name" {
+  type        = string
+  description = "PostgreSQL instance name. No special characters."
+}
+
+variable "db_username" {
+  type        = string
+  description = "PostgreSQL instance username. No special characters."
+}
```
Comment (nikolasrieble) [inline modules/aurora_database_cluster/variables.tf:17]: 🙏 Lets please use the validation feature if we have such requirements.
```suggestion
variable "db_name" {
  type        = string
  description = "PostgreSQL instance name. No special characters."
  
    validation {
    condition     = can(regex("^[a-zA-Z0-9]+$", var.db_name))
    error_message = "The db_name must only contain alphanumeric characters."
  }
}

variable "db_username" {
  type        = string
  description = "PostgreSQL instance username. No special characters."
  
  validation {
    condition     = can(regex("^[a-zA-Z0-9]+$", var.db_username))
    error_message = "The db_name must only contain alphanumeric characters."
  }
}
```

File: modules/networking/variables.tf [hunk 4]
```diff
@@ -16,7 +16,7 @@ variable "network_cidr" {
 variable "network_private_subnet_cidrs" {
   type        = list(string)
   description = "(Optional) List of private subnet CIDR ranges to create in VPC."
-  default     = ["10.0.32.0/20", "10.0.48.0/20"]
+  default     = ["10.0.32.0/20", "10.0.48.0/20", "10.0.112.0/20"]
```
Comment (nikolasrieble) [inline modules/networking/variables.tf:19]: 🙏 Lets please not change any defaults. It should be sufficient to set the values in the example instead. 
Changing the defaults is a breaking change, since it affects current users of this module. Is this change really necessary?

File: variables.tf [hunk 5]
```diff
@@ -165,6 +165,13 @@ variable "db_backup_retention" {
   default     = 0
 }
 
+variable "aurora_db_backup_retention" {
```
Comment (nikolasrieble) [inline variables.tf:168]: 🙏 Please group all the aurora settings next to each other.

**Review by nikolasrieble — COMMENTED** [review]
Thank you for your work! 💪 🚀 
I have a few questions and suggestions and will also test the module myself. 

In the meantime, I already wanted to share my feedback. 
Please also update the main README to refer to this example in the examples section.

File: modules/aurora_database_cluster/main.tf [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+# AZs
+#################################################
+data "aws_availability_zones" "available" {
+  state = "available"
+
+}
+
+resource "random_string" "postgresql_password" {
+  length  = 99
+  special = false
```
Comment (miguelhrocha) [inline modules/aurora_database_cluster/main.tf:14]: We want special characters 👍🏼

File: modules/aurora_database_cluster/main.tf [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+#################################################
+# AZs
+#################################################
+data "aws_availability_zones" "available" {
+  state = "available"
+
+}
+
+resource "random_string" "postgresql_password" {
+  length  = 99
```
Comment (miguelhrocha) [inline modules/aurora_database_cluster/main.tf:13]: Why such a large length? 🤔

**Review by miguelhrocha — COMMENTED** [review]
I have two additional comments 👍🏼

File: examples/standalone-rhel-aurora/provider.tf [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+provider "aws" {
+  assume_role {
+    role_arn = var.aws_role_arn
+  }
+
+  default_tags {
+    tags = local.common_tags
+  }
+}
```
Comment (nikolasrieble) [inline examples/standalone-rhel-aurora/provider.tf:13]: 🚧 We should not create a provider here. After I deployed the Aurora example with this branch, and wanted to tear it down again, I am facing the following issue: 

```
Error: Provider configuration not present
To work with module.tfe.module.standalone_rhel_aurora.module.aurora_database[0].aws_security_group_rule.postgresql_ingress (orphan) its original provider configuration at module.tfe.provider["registry.terraform.io/hashicorp/aws"] is required, but it has been removed. This occurs when a provider configuration is removed while objects created by that provider still exist in the state. Re-add the provider configuration to destroy module.tfe.module.standalone_rhel_aurora.module.aurora_database[0].aws_security_group_rule.postgresql_ingress (orphan), after which you can remove the provider configuration again.
```

If I understand correctly, we should require a provider instead of instantiating one. Please see how the other examples handle this, such as [standalone-airgap-dev](https://github.com/hashicorp/terraform-aws-terraform-enterprise/blob/main/examples/standalone-airgap-dev/versions.tf). 

Also, when I consult the [terraform docs](https://developer.hashicorp.com/terraform/language/modules/develop/providers), I find: 

> A module intended to be called by one or more other modules must not contain any provider blocks.

Please adapt this to follow the other examples.

**Review by nikolasrieble — CHANGES_REQUESTED** [review]

File: examples/standalone-rhel-aurora/outputs.tf [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+output "ptfe_endpoint" {
+  value       = module.standalone_rhel_aurora.tfe_url
+  description = "The URL to the TFE application."
+}
+
+output "tfe_url" {
+  value       = module.standalone_rhel_aurora.tfe_url
+  description = "The URL to the TFE application."
+}
```
[the remark this answers is not in the document]
Comment (sandrampeter) [reply examples/standalone-rhel-aurora/outputs.tf:12]: this was originally written for test where it is required, but removing as its an example

File: examples/standalone-rhel-aurora/locals.tf [hunk 10]
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) HashiCorp, Inc.
+# SPDX-License-Identifier: MPL-2.0
+
+locals {
+  common_tags = {
+    Environment = var.license_file == null ? "tfe_utilities_test" : "ptfe-replicated CI"
```
[the remark this answers is not in the document]
Comment (sandrampeter) [reply examples/standalone-rhel-aurora/locals.tf:6]: removed tags completely for aligning with example outputs

File: examples/standalone-rhel-aurora/outputs.tf [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+output "replicated_console_url" {
+  value       = "${module.standalone_rhel_aurora.tfe_url}:8800"
+  description = "Terraform Enterprise Console URL"
+}
+
+# Change this to health_check_url for consistency. This requires changing it in ptfe-replicated tests.
```
[the remark this answers is not in the document]
Comment (sandrampeter) [reply examples/standalone-rhel-aurora/outputs.tf:19]: done

File: modules/networking/variables.tf (hunk 4 above)
Comment (sandrampeter) [reply modules/networking/variables.tf:19]: reverted

File: modules/aurora_database_cluster/main.tf [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+  special = false
+}
+
+resource "aws_security_group" "postgresql" {
+  description = "The security group of the PostgreSQL deployment for TFE."
+  name        = "${var.friendly_name_prefix}-tfe-postgresql"
```
[the remark this answers is not in the document]
Comment (sandrampeter) [reply modules/aurora_database_cluster/main.tf]: done

File: modules/aurora_database_cluster/variables.tf (hunk 2 above)
Comment (sandrampeter) [reply modules/aurora_database_cluster/variables.tf]: I have renamed the vars to `aurora_cluster_instance_enable_single` : Creates only a single AWS RDS Aurora Cluster Instance and `aurora_cluster_instance_replica_count` : Number of extra cluster instances to create. Should be `0` if `aurora_cluster_instance_enable_single` is set to `true`.

File: modules/aurora_database_cluster/variables.tf (hunk 1 above)
Comment (sandrampeter) [reply modules/aurora_database_cluster/variables.tf]: renamed to `aurora_cluster_instance_replica_count` : Number of extra cluster instances to create. Should be `0` if `aurora_cluster_instance_enable_single` is set to `true`.

File: modules/aurora_database_cluster/main.tf (hunk 7 above)
Comment (sandrampeter) [reply modules/aurora_database_cluster/main.tf]: updated

File: modules/aurora_database_cluster/main.tf (hunk 6 above)
Comment (sandrampeter) [reply modules/aurora_database_cluster/main.tf]: updated

File: modules/aurora_database_cluster/main.tf [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+  port                         = 5432
+  preferred_backup_window      = var.db_backup_window
+  preferred_maintenance_window = var.preferred_maintenance_window
+
+
+  skip_final_snapshot = true
+  storage_encrypted   = true
+
+  vpc_security_group_ids = [aws_security_group.aurora_postgresql.id]
+}
```
Comment (miguelhrocha) [inline modules/aurora_database_cluster/main.tf:95]: This is a nit, but I'd like to hear your thoughts: I usually prefer to bundle related inputs in their block, for example:

1. Database setup i.e. user, pass would be 1 block
2. Encryption settings another block
3. Network settings another block

You get the gist. This reduces the mental context needed to understand a module/resource. Let me know what you think!

**Review by miguelhrocha — COMMENTED** [review]
Thank you for the iteration Sandra, have more comments 👍🏼

File: examples/standalone-aurora/variables.tf [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+  description = "The secrets manager secret ID of the Base64 & PEM encoded TLS private key for tfe."
+}
+
+variable "certificate_pem_secret_id" {
+  type        = string
+  description = "The secrets manager secret ID of the Base64 & PEM encoded TLS certificate for tfe."
+}
+
+variable "aurora_cluster_instance_enable_single" {
+  type        = string
```
[the remark this answers is not in the document]
Comment (sandrampeter) [reply examples/standalone-aurora/variables.tf:30]: done

File: modules/aurora_database_cluster/variables.tf [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+
+variable "kms_key_id" {
+  description = "The ARN for the KMS encryption key. When specifying `kms_key_id`, `storage_encrypted` needs to be set to `true`."
+  type        = string
+  default     = null
+}
+
+
+variable "preferred_maintenance_window" {
```
[the remark this answers is not in the document]
Comment (sandrampeter) [reply modules/aurora_database_cluster/variables.tf:93]: This is one of the var used for both database and aurora_database_cluster, which is why I couldnt add a prefix.

File: modules/aurora_database_cluster/outputs.tf [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+}
+
+output "name" {
+  value       = aws_rds_cluster.aurora_postgresql.database_name
+  description = "The name of the PostgreSQL RDS instance."
+}
+
+output "password" {
+  value       = random_string.aurora_postgresql_password.result
+  description = "The password of the main PostgreSQL user."
```
[the remark this answers is not in the document]
Comment (sandrampeter) [reply modules/aurora_database_cluster/outputs.tf:16]: just following the pattern from existing database module

File: modules/aurora_database_cluster/main.tf (hunk 13 above)
Comment (sandrampeter) [reply modules/aurora_database_cluster/main.tf:95]: This is already bundled alphabetically and placed together with related vars. This is similar to the ordering already in place.

**Review by nikolasrieble — APPROVED** [review]
Great work, thank you Sandra. 
Please make sure to once deploy this example and execute a few runs against it as a smoke test.
